sklearn_numba_dpex/kmeans/engine.py

```diff
@@ -156,155 +201,160 @@ def kmeans_single(self, X, sample_weight, centers_init):
+        with _validate_with_array_api():
+            init = check_array(
+                init,
+                dtype=X.dtype,
+                accept_sparse=False,
+                copy=False,
+                order=self.order,
+                force_all_finite=True,
+                ensure_2d=True,
+                estimator=self.estimator,
+                input_name="init",
             )
-            sample_weight = sample_weight.astype(compute_dtype)
+            self.estimator._validate_center_shape(X, init)
+            init_t = dpt.asarray(init.T, order="C", copy=False)
+            return init_t
 
-        return X, sample_weight, cluster_centers, output_dtype
 
-    def _load_transposed_data_to_device(self, X, sample_weight, cluster_centers):
-        # Transfer the input data to device memory,
-        # TODO: let the user pass directly dpt or dpnp arrays to avoid copies.
+def _get_namespace(*arrays):
+    return dpt, True
 
-        # NB: numba_dpex kernels only currently supports inputs with a C memory layout
-        # (see https://github.com/IntelPython/numba-dpex/issues/767) but our KMeans
-        # implementation is hypothetized to be more efficient with the F-memory layout.
-        # As a workaround the kernels work with the transpose of X, X_t, where X_t
-        # is created with a C layout, which results in equivalent memory access
-        # patterns than with a F layout for X.
-        # TODO: when numba_dpex supports inputs with F-layout:
-        # - use X rather than X_t and adapt the codebase (better for readability and
-        # more consistent with sklearn notations)
-        # - test the performances with both layouts and use the best performing layout.
 
-        X_t = dpt.asarray(X.T, order="C", device=self.device)
-        assert (
-            X_t.strides[1] == 1
-        )  # C memory layout, equivalent to Fortran layout on transposed
+def _asarray_with_order(array, dtype, order, copy=None, xp=None):
+    return dpt.asarray(array, dtype=dtype, order=order, copy=copy)
 
-        if sample_weight is not None:
-            sample_weight = dpt.from_numpy(sample_weight, device=self.device)
 
-        if cluster_centers is not None:
-            cluster_centers = dpt.from_numpy(cluster_centers.T, device=self.device)
-
-        return X_t, sample_weight, cluster_centers
+@contextlib.contextmanager
+def _validate_with_array_api():
+    # TODO: when https://github.com/IntelPython/dpctl/issues/997 and
+    # https://github.com/scikit-learn/scikit-learn/issues/25000 and are solved
+    # remove those hacks.
+    with sklearn.config_context(
+        array_api_dispatch=True,
+        assume_finite=True  # workaround 1: disable force_all_finite
+        # workaround 2: monkey patch get_namespace and _asarray_with_order to force
+        # dpctl.tensor array namespace
+    ), override_attr_context(
+        sklearn_validation,
+        get_namespace=_get_namespace,
+        _asarray_with_order=_asarray_with_order,
```

Comment:
Using monkey patching here enables a very efficient workflow as a whole. I wonder if rather than having those 2 functions limited to array api specs, we should also consider pluginying it.